core/market-decentralized/src/market.rs

```diff
@@ -0,0 +1,122 @@
+use ya_core_model::market::BUS_ID;
+use ya_persistence::executor::DbExecutor;
+use ya_service_api_interfaces::{Provider, Service};
+
+#[derive(Error, Debug)]
+pub enum MarketError {}
+
+#[derive(Error, Debug)]
+pub enum MarketInitError {
+    #[error("Failed to initialize Offers matcher. Error: {}.", .0)]
+    Matcher(#[from] MatcherInitError),
+    #[error("Failed to initialize negotiation engine. Error: {}.", .0)]
+    Negotiation(#[from] NegotiationInitError),
+}
+
+/// Structure connecting all market objects.
+pub struct Market {
+    matcher: Arc<Matcher>,
+    provider_negotiation_engine: Arc<ProviderNegotiationEngine>,
+    requestor_negotiation_engine: Arc<RequestorNegotiationEngine>,
+}
+
+impl Market {
+    pub fn new(db: &DbExecutor, data_dir: &Path) -> Result<Self, MarketInitError> {
+        // TODO: Set Matcher independent parameters here or remove this todo.
+        let builder = DiscoveryBuilder::new();
+
+        let (matcher, listeners) = Matcher::new::<DiscoveryGSB>(builder, data_dir)?;
+        let provider_engine = ProviderNegotiationEngine::new(db.clone())?;
+        let requestor_engine =
+            RequestorNegotiationEngine::new(db.clone(), listeners.proposal_receiver)?;
+
+        Ok(Market {
+            matcher: Arc::new(matcher),
+            provider_negotiation_engine: provider_engine,
+            requestor_negotiation_engine: requestor_engine,
+        })
+    }
+
+    pub async fn bind_gsb(&self, prefix: String) -> Result<(), MarketInitError> {
+        self.matcher.bind_gsb(prefix.clone()).await?;
+        self.provider_negotiation_engine
+            .bind_gsb(prefix.clone())
+            .await?;
+        self.requestor_negotiation_engine.bind_gsb(prefix).await?;
+        Ok(())
+    }
+
+    pub async fn gsb<Context: Provider<Self, DbExecutor> + Provider<Self, PathBuf>>(
+        ctx: &Context,
+    ) -> anyhow::Result<()> {
+        let db = ctx.component();
+        let data_dir: PathBuf = ctx.component();
+        let market = MARKET.get_or_init_market(&db, &data_dir)?;
+        Ok(market.bind_gsb(BUS_ID.to_string()).await?)
+    }
+
+    pub fn rest(db: &DbExecutor) -> actix_web::Scope {
+        // TODO: We should run gsb function first, because it gets datadir.
+        //       This code is order dependent :/
+        let market = match MARKET.get_or_init_market(db, &current_dir().unwrap()) {
+            Ok(market) => market,
+            Err(error) => {
+                log::error!("{}", error);
+                panic!("Market initialization impossible. Check error logs.")
+            }
+        };
+
+        actix_web::web::scope(crate::MARKET_API_PATH).data(market)
+    }
+}
+
+impl Service for Market {
+    type Cli = ();
+}
+
+// =========================================== //
+// Awful static initialization. Necessary to
+// share Market between gsb and rest functions.
+// =========================================== //
+
+struct StaticMarket {
+    locked_market: Mutex<Option<Arc<Market>>>,
```

Comment:
`Arc<Mutex>` is much more usable than opposite `Mutex<Arc>`

Reply:
Note that this mutex is protecting not Market, but Option, which will change from None to Some, so it must be mutable. Market is not designed to be wrapped in mutex. All functions on Market should get &self. Synchronization must be done in internal structures (if necessary, but we have db, so maybe it won't be).

If we would do, as you suggested, we had to lock each market call. It is impossible to work with async calls than.

Reply:
let's discuss this on a call